localstack/aws/protocol/parser.py

```diff
@@ -248,7 +248,17 @@ def _parse_shape(
             elif location == "uri":
                 uri_param_name = shape.serialization.get("name")
                 if uri_param_name in uri_params:
-                    payload = uri_params[uri_param_name]
+                    # FIXME special case for s3 object-names (=key): trailing '/' are valid and need to be preserved
+                    #       however, the url-matcher removes it from the key
+                    #       we check the request.url to verify the name
+                    if (
+                        self.service.service_name == "s3"
+                        and uri_param_name == "Key"
+                        and request.url.split("?")[0].endswith(f"{uri_params[uri_param_name]}/")
```

Comment:
predictably i'd prefer encapsulating this in the s3 parser somehow :D i don't have a good suggestion at the moment, will take a look after the review, maybe @alexrashed also has some ideas

Reply:
To be host, the `service_name` is something we desperately tried to avoid in these "generalized" parsers.
Since there is already an `S3RequestParser`, we could check this case and modifying the `uri_params` and afterwards calling the super method. Something like this:
```python
class S3RequestParser(RestXMLRequestParser):
    def _parse_shape(
        self, request: HttpRequest, shape: Shape, node: Any, uri_params: Mapping[str, Any] = None
    ) -> Any:
        if shape is not None and uri_params is not None \
                and shape.serialization.get("location") == "uri" \
                and shape.serialization.get("name") == "Key" \
                and request.url.split("?")[0].endswith(f"{uri_params['Key']}/"):
            uri_params = dict(uri_params)
            uri_params["Key"] = uri_params["Key"] + "/"
        return super()._parse_shape(request, shape, node, uri_params)
```

Another option would be to externalize the location parsing to a separate function, but this doesn't seem to be worth the effort / complexity.

I'm happy to assist if I can be of service!